### DISCUSSION TITLE
Enhanced interactive ROHD simulation debug

### DISCUSSION THREAD

**Discussion by mkorbel1** [intel/rohd#125] [Ideas]
I've been thinking about ways to make interactive debug with a ROHD design easier.  With SystemVerilog designs, interactive debug can be pretty painful, but there are (proprietary) tools available that allow tracing of signals through original source code and relation to waveforms.

With ROHD, we already have a really great interactive debug ability through the Dart debugger and its attachment to IDEs like vscode.  However, the debugger has a couple limitations that are a little annoying:
- The waveform generated as a .vcd file is completely unconnected from the debug process
- The debugger reveals all internal variables of objects with equal priority, but when debugging hardware you're most interested in a few variables like current value, source/destination connections, etc.

I think it would be really cool to build a tool that connects to the Dart debugger and cherry-picks hardware-debug-related information.  Waveform history per-signal could be stored in debug-visible variables rather than VCD, which would mean we could attach a live waveform view to that same debug process.

These links seem potentially relevant/interesting:
https://code.visualstudio.com/api/extension-guides/debugger-extension
https://microsoft.github.io/debug-adapter-protocol/

This is probably not a small project to take on, but could really elevate the debuggability of ROHD.

**Comment by quekyj** [comment]
https://github.com/Dart-Code/Dart-Code/issues/4589

We will start to work on the debugging tools for ROHD. We are going to start small by having a `Debug Adapter` that can just filter signals out from Dart Developer Service [`dds` package](https://pub.dev/packages/dds). Anyone are welcome to work on this project together by commenting here!

**Reply by quekyj** [reply]
The proposal on the debugger dev tool can be find at https://github.com/intel/rohd/discussions/418
